packages/graphiql/__mocks__/@graphiql/react.ts

```diff
@@ -0,0 +1,98 @@
+import {
+  EditorContext,
+  EditorContextProvider,
+  ExplorerContext,
+  ExplorerContextProvider,
+  useExplorerNavStack,
+  useHeaderEditor as _useHeaderEditor,
+  useQueryEditor as _useQueryEditor,
+} from '@graphiql/react';
+import type {
+  EditorContextType,
+  ExplorerContextType,
+  ExplorerFieldDef,
+  ExplorerNavStack,
+  ExplorerNavStackItem,
+  UseHeaderEditorArgs,
+  UseQueryEditorArgs,
+} from '@graphiql/react';
+import { useEffect, useRef, useState } from 'react';
+
+export {
+  EditorContext,
+  EditorContextProvider,
+  ExplorerContext,
+  ExplorerContextProvider,
+  useExplorerNavStack,
+};
+
+export type {
+  EditorContextType,
+  ExplorerContextType,
+  ExplorerFieldDef,
+  ExplorerNavStack,
+  ExplorerNavStackItem,
+  UseHeaderEditorArgs,
+  UseQueryEditorArgs,
+};
+
+function useMockedEditor(value?: string, onEdit?: (newValue: string) => void) {
+  const [code, setCode] = useState(value);
+  const ref = useRef<HTMLDivElement>(null);
+
+  useEffect(() => {
+    if (!ref.current) {
+      return;
+    }
+
+    const mockGutter = document.createElement('div');
+    mockGutter.className = 'CodeMirror-gutter';
+
+    const mockTextArea = document.createElement('textarea');
+    mockTextArea.className = 'mockCodeMirror';
+    mockTextArea.addEventListener('change', event => {
+      const newValue = (event.target as HTMLTextAreaElement).value;
+      setCode(newValue);
+      onEdit?.(newValue);
+    });
+
+    const mockWrapper = document.createElement('div');
+    mockWrapper.appendChild(mockGutter);
+    mockWrapper.appendChild(mockTextArea);
+
+    ref.current.appendChild(mockWrapper);
+  });
```

Comment:
Are you sure you don't want to add a `[]` dependency here? With this, `useEffect` will be called on every render.

Reply:
Good call 👍

Reply:
This effect is now still executing on each render, but we return early if the ref is set and already contains children. That way we also make sure that we insert the mock text area if the ref is not yet set on the initial render.